### PR TITLE
gitlab_project_access_token: Test bugfix and destroy improvement

### LINKED ISSUE #1002: gitlab_project_access_token - Could not provision maintainer access to project access token
### GitLab Provider version

_No response_

### GitLab version

_No response_

### Terraform version

_No response_

### Relevant Terraform Configuration

```hcl
resource "gitlab_project" "foo" {
  name = "foo-123"
  description = "Terraform acceptance tests"

  # So that acceptance tests can be run in a gitlab organization
  # with no billing
  visibility_level = "public"
}

resource "gitlab_project_access_token" "bar" {
  name = "my project token"
  project = gitlab_project.foo.id
  expires_at = "2022-04-01"
  scopes = ["read_repository" , "api", "write_repository", "read_api"]
}
```

### Relevant log output

```shell
=== RUN   TestAccGitlabProjectAccessToken_basic
166
    resource_gitlab_project_access_token_test.go:22: Step 1/6 error: Error running apply: exit status 1
167
        
168
        Error: POST http://127.0.0.1:8080/api/v4/projects/42/access_tokens: 400 {message: 400 Bad request - Could not provision maintainer access to project access token}
169
        
170
          with gitlab_project_access_token.bar,
171
          on terraform_plugin_test.tf line 11, in resource "gitlab_project_access_token" "bar":
172
          11: resource "gitlab_project_access_token" "bar" {
173
        
174
--- FAIL: TestAccGitlabProjectAccessToken_basic (6.45s)
175
=== RUN   TestAccGitlabProjectAccessToken_accessLevel
176
    resource_gitlab_project_access_token_test.go:102: Step 1/3 error: Error running apply: exit status 1
177
        
178
        Error: POST http://127.0.0.1:8080/api/v4/projects/44/access_tokens: 400 {message: 400 Bad request - Could not provision maintainer access to project access token}
179
        
180
          with gitlab_project_access_token.bar,
181
          on terraform_plugin_test.tf line 11, in resource "gitlab_project_access_token" "bar":
182
          11: resource "gitlab_project_access_token" "bar" {
183
        
184
--- FAIL: TestAccGitlabProjectAccessToken_accessLevel (6.48s)
```

### Description

The pipelines have been failing since merging #997 

See: https://github.com/gitlabhq/terraform-provider-gitlab/runs/5796647075?check_suite_focus=true

### PR DESCRIPTION
**Pull request by armsnyder**
## Description

Fixes #1002 

- Fixed a test bug where a hardcoded expiration date that passed caused failures
- Improved the destroy function to wait for the token to finish deleting

Note the diff to the `resourceGitlabProjectAccessTokenRead` looks large, but actually it is a straight move. I moved the list API calls into a `resourceGitlabProjectAccessTokenFind` function because I had to use it in the destroy logic also.

### PR Checklist Acknowledgement

- [x] I acknowledge that all of the following items are true, where applicable:
  - Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/).
  - [Examples](https://github.com/gitlabhq/terraform-provider-gitlab/tree/main/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
  - New resources have at minimum a basic test with three steps:
    - Create the resource
    - Update the attributes
    - Import the resource
  - No new `//lintignore` comments were copied from existing code. (Linter rules are meant to be enforced on new code.)

### CODE REVIEW

File: internal/provider/resource_gitlab_project_access_token_test.go [hunk 1]
```diff
@@ -2,378 +2,103 @@ package provider
...
-						name:        "my project token",
-						scopes:      map[string]bool{"read_repository": true, "api": true, "write_repository": true, "read_api": true},
-						expiresAt:   "2022-04-01",
-						accessLevel: accessLevelValueToName[gitlab.DeveloperPermissions],
-					}),
+					resource.TestCheckResourceAttr("gitlab_project_access_token.foo", "active", "true"),
+					resource.TestCheckResourceAttr("gitlab_project_access_token.foo", "revoked", "false"),
+					resource.TestCheckResourceAttrSet("gitlab_project_access_token.foo", "token"),
+					resource.TestCheckResourceAttrSet("gitlab_project_access_token.foo", "created_at"),
+					resource.TestCheckResourceAttrSet("gitlab_project_access_token.foo", "user_id"),
```
Comment (PatrickRice-KSC) [inline internal/provider/resource_gitlab_project_access_token_test.go:67]: @armsnyder  - Given we're setting a "expires_at" with a specific value (and this is the only test that does so), it seems like we may want to test that the value matches what we pass in. I know there is an import test right after, but it seems an explicit check here may be good.

Comment (PatrickRice-KSC) [reply internal/provider/resource_gitlab_project_access_token_test.go:67]: Same, likely, with the access level (since those explicit tests were removed)

**Comment by PatrickRice-KSC** [comment]
Very minor comment, otherwise LGTM. I think we should prioritize getting this merged so other PRs can rebase and ensure their checks pass, so let me know if there is anything I can do to help!

Comment for others - It makes sense to me that the "update" tests were largely removed, given pretty much every attribute of the access_token is a "ForceNew" attribute, and some of the newer tests were just wrapped into the second step of the "Basic" test. So even though the code looks like a bunch of lines were removed, I think this simplification covers the same general test cases. Given it's simpler, to me that means it's a good change!

File: internal/provider/resource_gitlab_project_access_token_test.go (hunk 1 above)
Comment (armsnyder) [reply internal/provider/resource_gitlab_project_access_token_test.go:67]: After each TestStep, the framework re-runs `terraform plan` and fails the test if the plan is not empty, meaning that the values in state match what you specified in the config. Please correct me if I'm wrong, but I don't think there is a point to using `resource.TestCheckResourceAttr` functions to check what you set in the config.

**Comment by armsnyder** [comment]
> It makes sense to me that the "update" tests were largely removed, given pretty much every attribute of the access_token is a "ForceNew" attribute

Yeah, and to add more context, since this resource has no "update" function at all, the Terraform SDK enforces that all attributes have `ForceNew: true`.

File: internal/provider/resource_gitlab_project_access_token_test.go (hunk 1 above)
Comment (PatrickRice-KSC) [reply internal/provider/resource_gitlab_project_access_token_test.go:67]: @armsnyder - Ah, I knew it did that in the import steps, I didn't realize it did that automatically after every step, but I suppose that's why the `ExpectNonEmptyPlan` exists logically. 

Seems reasonable, I'm good.

**Comment by PatrickRice-KSC** [comment]
> > It makes sense to me that the "update" tests were largely removed, given pretty much every attribute of the access_token is a "ForceNew" attribute
> 
> Yeah, and to add more context, since this resource has no "update" function at all, the Terraform SDK enforces that all attributes have `ForceNew: true`.

Learn new things every day, thanks for the info!

**Review by PatrickRice-KSC — APPROVED** [review]

**Comment by PatrickRice-KSC** [comment]
@nagyv  - This is one that would be a great priority to merge, if you wouldn't mind, so we can have other PRs rebase and get their checks passing!

File: internal/provider/resource_gitlab_project_access_token_test.go (hunk 1 above)
Comment (armsnyder) [reply internal/provider/resource_gitlab_project_access_token_test.go:67]: Oddly, in the link I shared, their examples do use `resource.TestCheckResourceAttr` to check the attributes set in the config. This is probably because there are edge cases where you might have a `DiffSuppressFunc` that renders the empty-plan check not useful.

My feeling is still that the combination of the test framework's features + using `ImportStateVerify` means we don't need to check these, since `ImportStateVerify` uses strict equality without respecting `DiffSuppressFunc` or `CustomizeDiff`, but I plan on opening an issue so we can discuss more.

Resolving this for now through.

**Review by timofurrer — APPROVED** [review]

**Comment by timofurrer** [comment]
oh boy, what a fail 🤦 Thanks for fixing this @armsnyder 🎉
